src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -582,57 +586,69 @@ private void RestoreUncommittedInput()
+                        }
+                        code = sb.ToString();
+                    }
+                    else if (_window.InteractiveWindowClipboard.ContainsText())
+                    {
+                        code = _window.InteractiveWindowClipboard.GetText();
                     }
                     else
                     {
                         return false;
                     }
-                }
-                else if (IsInActivePrompt(TextView.Caret.Position.BufferPosition))
-                {
-                    MoveCaretToClosestEditableBuffer();
-                }
-                else if (MapToEditableBuffer(TextView.Caret.Position.BufferPosition) == null)
-                {
-                    return false;
-                } 
 
-                string format = Evaluator.FormatClipboard();
-                if (format != null)
-                {
-                    InsertCode(format);
-                }
-                else if (_window.InteractiveWindowClipboard.ContainsData(ClipboardFormat))
-                {
-                    var blocks = BufferBlock.Deserialize((string)_window.InteractiveWindowClipboard.GetData(ClipboardFormat));
-                    // Paste each block separately.
-                    foreach (var block in blocks)
+                    // Delete selected text if there's any and adjust caret position
+                    if (!TextView.Selection.IsEmpty)
                     {
-                        switch (block.Kind)
+                        if (CutOrDeleteSelection(isCut: false))
+                        {
+                            MoveCaretToClosestEditableBuffer();
+                        }
+                        else
                         {
-                            case ReplSpanKind.Input:
-                            case ReplSpanKind.Output:
-                            case ReplSpanKind.StandardInput:
-                                InsertCode(block.Content);
-                                break;
+                            return false;
                         }
                     }
-                }
-                else if (_window.InteractiveWindowClipboard.ContainsText())
-                {
-                    InsertCode(_window.InteractiveWindowClipboard.GetText());
-                }
-                else
-                {
-                    return false;
-                }
+                    else if (IsInActivePrompt(TextView.Caret.Position.BufferPosition))
+                    {
+                        MoveCaretToClosestEditableBuffer();
+                    }
+                    else if (MapToEditableBuffer(TextView.Caret.Position.BufferPosition) == null)
+                    {
+                        return false;
+                    }
 
-                return true;
+                    // do the paste and complete the transection
+                    InsertCode(code);
+                    transaction.Complete();
```

Comment:
This actually doesn't work. I suspect it is because the edit in `CutOrDelete` is done on a different text buffer, which is `CurrentLanguageBuffer` instead of `TextView.TextBuffer`.